src/services/factories/tokenList.ts

```diff
@@ -12,58 +12,103 @@ export function getTokensFactory(factoryParams: {
 
   const updatedIdsForNetwork = new Set<number>()
 
-  interface GetTokenIdResult {
-    value: number | null
+  interface UpdateTokenIdResult {
+    token: TokenDetails
     remove?: true
+    failed?: true
   }
 
-  async function getTokenId(networkId: number, tokenAddress: string): Promise<GetTokenIdResult> {
+  async function updateTokenId(networkId: number, token: TokenDetails): Promise<UpdateTokenIdResult> {
     try {
-      const hasToken = await exchangeApi.hasToken({ networkId, tokenAddress })
+      const hasToken = await exchangeApi.hasToken({ networkId, tokenAddress: token.address })
       if (!hasToken) {
-        return { value: null, remove: true }
+        return { token, remove: true }
       }
-      return { value: await exchangeApi.getTokenIdByAddress({ networkId, tokenAddress }) }
+      token.id = await exchangeApi.getTokenIdByAddress({ networkId, tokenAddress: token.address })
+      return { token }
     } catch (e) {
-      logDebug(`[network:${networkId}][address:${tokenAddress}] Failed to fetch id from contract`, e.message)
-      return { value: null }
+      logDebug(`[network:${networkId}][address:${token.address}] Failed to fetch id from contract`, e.message)
+      return { token, failed: true }
     }
   }
 
-  async function updateTokenIds(networkId: number): Promise<void> {
-    // Set as updated on start to prevent concurrent queries
-    updatedIdsForNetwork.add(networkId)
+  async function updateWithRetries(
+    networkId: number,
+    tokens: TokenDetails[],
+    maxRetries = 3,
+    retryDelay = 1000,
+    exponentialBackOff = true,
+  ): Promise<void> {
+    if (maxRetries <= 0) {
+      throw new Error(
+        `[network:${networkId}] Max retries exceeded trying to fetch token ids for tokens ${tokens.map(
+          t => t.address,
+        )}`,
+      )
+    }
 
-    const promises = tokenListApi.getTokens(networkId).map(token => getTokenId(networkId, token.address))
+    const promises = tokens.map(token => updateTokenId(networkId, token))
 
     const results = await Promise.all(promises)
 
-    let failedToUpdate = false
-
-    const tokenList = results.reduce((tokens: TokenDetails[], result: GetTokenIdResult, index: number) => {
-      if (result.value !== null) {
-        // We got a result, yay
-        const token = tokenListApi.getTokens(networkId)[index]
-        token.id = result.value
-        tokens.push(token)
-      } else if (!result.remove) {
-        // Failed to query, but don't remove from the list
-        tokens.push(tokenListApi.getTokens(networkId)[index])
-        // Try again next time
-        // TODO: try again only for the ones that failed
-        failedToUpdate = true
+    const updated = new Map<string, TokenDetails>()
+    const failedToUpdate: TokenDetails[] = []
+    const toRemove = new Set<string>()
+
+    // classify results into map/list/set for easier access
+    results.forEach(({ token, remove, failed }) => {
```

Comment:
Maybe I'm mistaken, but if `result` is the result of Promise.all, then in case of an error it will throw already in line `52` and you are not catching that error. The retry logic depends on `failedToUpdate` that is initialized after

Reply:
edit: All right, I see, `updateTokenId` in his implementation make sure we never throw. so I guess is just a nitpick to make it more clear, so you don't need to check the implementation.
Either:
* making it more evident in the name `updateTokenIdOrXXXXXX` (being XXXXXX whatever you think clarifies here, too mentally tired now)
* ...or, make the catch inline `tokens.map(token => updateTokenId(networkId, token).catch(_handleFailedToken))`

Reply:
Updated the function name.